### semana 2 

In [6]:
import numpy as np
print (np.__version__)

import pandas as pd
print (pd.__version__)

import matplotlib as plt
print (plt.__version__)

import ipyparallel as ipp
print (ipp.__version__)

1.11.3
0.19.2
2.0.0
6.0.2


In [8]:
rc = ipp.Client()

In [9]:
rc.ids

[0, 1]

In [10]:
rc

In [11]:
# %load ~/.ipython/profile_default/security/ipcontroller-engine.json
{
  "ssh": "",
  "interface": "tcp://127.0.0.1",
  "registration": 51400,
  "control": 51407,
  "mux": 51409,
  "hb_ping": 51410,
  "hb_pong": 51411,
  "task": 51413,
  "iopub": 51415,
  "key": "daedf0c7-4cce-4bbb-bc3d-fc3fd7d8fef7",
  "location": "192.168.10.201",
  "pack": "json",
  "unpack": "json",
  "signature_scheme": "hmac-sha256"
}

{'control': 51407,
 'hb_ping': 51410,
 'hb_pong': 51411,
 'interface': 'tcp://127.0.0.1',
 'iopub': 51415,
 'key': 'daedf0c7-4cce-4bbb-bc3d-fc3fd7d8fef7',
 'location': '192.168.10.201',
 'mux': 51409,
 'pack': 'json',
 'registration': 51400,
 'signature_scheme': 'hmac-sha256',
 'ssh': '',
 'task': 51413,
 'unpack': 'json'}

#### Uso DirectView

In [13]:
rc[:]

<DirectView [0, 1]>

In [15]:
# utiliza todos los motores que se definieron inicialmente
rc[:].apply_sync(lambda : "Hola alumnos")

['Hola alumnos', 'Hola alumnos']

In [14]:
# asigna todos los motores a la variable dview
dview = rc[:] 
dview.apply_sync(lambda : "Hola alumnos")

['Hola alumnos', 'Hola alumnos']

In [18]:
#lista de los motores
dview

<DirectView all>

In [19]:
#y su equivalente dview = c[:] 
dview = rc.direct_view()

In [20]:
import os 
dview.apply_sync(os.getpid)

[1292, 568]

In [21]:
import socket
dview.apply_sync(socket.gethostname)

['Lupitha-PC', 'Lupitha-PC']

In [22]:
#suma de numeros en rango de 10
dview.map_sync(lambda x, y, z: x+y+z, range(10), range(10), range(10))

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27]

In [24]:
rc[::3].map_sync(lambda x, y, z: x + y + z, range(10), range(10), range(10))

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27]

In [26]:
#se usa un motor fijo
rc[1].map_sync(lambda x, y, z: x + y + z, range(10), range(10), range(10))

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27]

#### lview --> LoadBalancedView

In [28]:
lview = rc.load_balanced_view()

In [30]:
lview.map_sync(lambda x: sum(x), np.random.random((10,10000)))

[5014.6380767408973,
 4981.0941173064939,
 4998.6882389524935,
 4968.2511034724675,
 5000.5352140339846,
 5029.3142335267012,
 5042.4351579360145,
 5053.6064520867576,
 5038.8485597076815,
 4997.1390918466113]

In [32]:
def cubo(x):
    return x**3
cubo(4)

64

In [35]:
for x in range(1,7):
    print (cubo (x))

1
8
27
64
125
216


In [36]:
lista = [cubo(x)for x in range(1,7)]
lista

[1, 8, 27, 64, 125, 216]

#### Funcion map

In [38]:
map (cubo, range(1,7))

In [40]:
serial = map(cubo, range(1,7))

In [41]:
serial

In [43]:
parallel = dview.map_sync(cubo, range(1,70))

In [44]:
parallel

[1,
 8,
 27,
 64,
 125,
 216,
 343,
 512,
 729,
 1000,
 1331,
 1728,
 2197,
 2744,
 3375,
 4096,
 4913,
 5832,
 6859,
 8000,
 9261,
 10648,
 12167,
 13824,
 15625,
 17576,
 19683,
 21952,
 24389,
 27000,
 29791,
 32768,
 35937,
 39304,
 42875,
 46656,
 50653,
 54872,
 59319,
 64000,
 68921,
 74088,
 79507,
 85184,
 91125,
 97336,
 103823,
 110592,
 117649,
 125000,
 132651,
 140608,
 148877,
 157464,
 166375,
 175616,
 185193,
 195112,
 205379,
 216000,
 226981,
 238328,
 250047,
 262144,
 274625,
 287496,
 300763,
 314432,
 328509]

#### Funcion timeit --> tiempo en ejecutar cierto comando

In [48]:
%timeit map(cubo, range(1,40))
%timeit dview.map_sync(cubo, range(1,40))

1000000 loops, best of 3: 1.23 µs per loop
10 loops, best of 3: 51.3 ms per loop


#### Funcion lamba
##### Una función lambda es una función que toma cualquier número de argumentos y devuelve el valor de una expresión simple

In [49]:
PrimeQ = lambda x: all(x % i != 0 for i in range(int(x**0.5)+1)[2:]) and x>1

In [55]:
PrimeQ(23)

True

### Actividad



In [56]:
import numpy as np
import pandas as pd
import ipyparallel as ipp
import matplotlib as plt

Determinar si un numero de la forma $2^n -1$, donde $n$ es un entero, es primo o no. Los primos que tienen esta forma se llaman primos de Mersenne.
* Escribir una funcion que determine si el numero dado por $2^n -1$ es primo (usar la funcion anterior)
* Seleccionar los primos de Mersenne de entre los primeros 5000 numeros enteros (nota: el resultado es {2, 3, 5, 7, 13, 17, 19, 31, 61, 89, 107, 127, 521, 607, 1279, 2203, 2281, 3217, 4253, 4423}
* Estimar el tiempo que su computadora tarda en encontrar los primero 5000 primos de Mersenne


In [57]:
PrimeQ = lambda x: all(x % i != 0 for i in range(int(x**0.5)+1)[2:]) and x>1
ListaPrimos = [PrimeQ(2**n-1) for n in range(50)]

In [58]:
for x in range (49):
    if (ListaPrimos[x] == True):
        print (x)

2
3
5
7
13
17
19
31


In [60]:
for x in range (49):
    if (ListaPrimos[x] == True):
        print (2**x-1)

3
7
31
127
8191
131071
524287
2147483647


In [ ]:
%timeit ListaPrimos = [PrimeQ(2**n-1) for n in range(80)]